### ISSUE TITLE
V6 Cryptography - Key wrapping explanation

### ISSUE THREAD

**Issue by danielcuthbert** [OWASP/ASVS#2246]
<img width="934" alt="Screenshot 2024-11-05 at 14 19 45" src="https://github.com/user-attachments/assets/c7c1fda2-7c40-4436-88b7-71a4aa1f3370">

**Comment by danielcuthbert** [comment]
I totally agree with this and feel we should not only explain what key wrapping is, but also when it should be used. My attempt at this is as follows (please by all means give feedback on if you feel it is legitimate and usable)

> Cryptographic key wrapping is a method of protecting keys by encrypting a key using symmetric encryption in order to encapsulate said keys/material. The key used to encrypt the other key is called the wrapping key. This is often performed when you want to protect keys in places deemed untrustworthy, or send sensitive keys over untrusted networks or within applications.

**Comment by james-bitherder** [comment]
Suggested edit:
Cryptographic key wrap (and corresponding key unwrap) is a method of protecting an existing key by encapsulating (i.e. wrapping) it by employing an additional encryption mechanism so that the original key is not obviously exposed, e.g. during a transfer. This additional key used to protect the original key is referred to as the the wrap key.
This operation may be performed when it is desirable to protect keys in places deemed untrustworthy, or to send sensitive keys over untrusted networks or within applications.
However, serious consideration should be given to understanding the nature (e.g. the identity and the purpose) of the original key prior to committing to a wrap/unwrap procedure as this may have repercussions for both source and target systems/applications in terms of security and especially compliance which may include audit trails of a key's function (e.g. signing) as well as appropriate key storage.

**Comment by randomstuff** [comment]
I so far do not really know/understand (but I shall do my homework and find the answer :wink:) why key wrapping as a special construct is useful in contrast to encrypting the key with some general encryption algorithm. Maybe this should be clarified here?

**Comment by danielcuthbert** [comment]
It should, I'm just not sure if this is an Appendix addition over a requirement.

**Comment by jmanico** [comment]
...is key wrapping the same as envelope encryption?

**Comment by randomstuff** [comment]
So I tried to do my homework and AFAIU, AES-Keywrap is just some AES-based authenticated encryption.

[RFC 3394](https://datatracker.ietf.org/doc/html/rfc3394):

> Throughout this document, any data being wrapped will be referred to as the key data.  It makes no difference to the algorithm whether the data being wrapped is a key; in fact there is often good reason to include other data with the key, to wrap multiple keys together, or to wrap data that isn't strictly a key.  So, the term "key data" is used broadly to mean any data being wrapped, but particularly keys, since this is primarily a key wrap algorithm. 
>
> [...]
>
> The key wrap algorithm includes a strong integrity check on the key data.  If unwrapping produces the expected check value in A[0], then the chance that the key data is corrupt is 2^-64.

See as well "[When do I use NIST AES key wrapping?](https://security.stackexchange.com/questions/40052/when-do-i-use-nist-aes-key-wrapping)]"
